relay-server/src/actors/processor.rs

```diff
@@ -951,77 +962,333 @@ impl EnvelopeProcessorService {
+            return Some(ItemType::Attachment);
         }
 
-        session::process(state, &self.inner.config);
+        items.clone().next().map(|item| item.ty().clone())
+    }
+
+    fn process_event(&self, state: &mut ProcessEnvelopeState) -> Result<(), ProcessingError> {
+        // Events can also contain user reports.
         report::process(
             state,
             &self.inner.config,
             self.inner.outcome_aggregator.clone(),
         );
-        replay::process(state, &self.inner.config)?;
+
+        event::extract(state, &self.inner.config)?;
+
+        if_processing!(self.inner.config, {
+            attachment::create_placeholders(state);
+        });
+
+        event::finalize(state, &self.inner.config)?;
+        self.light_normalize_event(state)?;
+        dynamic_sampling::normalize(state);
+        event::filter(state)?;
+        dynamic_sampling::run(state, &self.inner.config);
+        dynamic_sampling::sample_envelope(state)?;
+
+        if_processing!(self.inner.config, {
+            event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
+            self.enforce_quotas(state)?;
+        });
+
+        if state.has_event() {
+            event::scrub(state)?;
+            event::serialize(state)?;
+        }
+
+        attachment::scrub(state);
+
+        Ok(())
+    }
+
+    fn process_transaction(&self, state: &mut ProcessEnvelopeState) -> Result<(), ProcessingError> {
         profile::filter(state);
-        span::filter(state);
+        event::extract(state, &self.inner.config)?;
+        profile::transfer_id(state);
 
-        if state.creates_event() {
-            // Some envelopes only create events in processing relays; for example, unreal events.
-            // This makes it possible to get in this code block while not really having an event in
-            // the envelope.
+        if_processing!(self.inner.config, {
+            attachment::create_placeholders(state);
+        });
+
+        event::finalize(state, &self.inner.config)?;
+        self.light_normalize_event(state)?;
+        dynamic_sampling::normalize(state);
+        event::filter(state)?;
+        dynamic_sampling::run(state, &self.inner.config);
+
+        // We avoid extracting metrics if we are not sampling the event while in non-processing
+        // relays, in order to synchronize rate limits on indexed and processed transactions.
+        if self.inner.config.processing_enabled() || state.sampling_result.should_drop() {
+            self.extract_metrics(state)?;
+        }
+
+        dynamic_sampling::sample_envelope(state)?;
 
-            if_processing!({
-                unreal::expand(state, &self.inner.config)?;
+        if_processing!(self.inner.config, {
+            event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
+            self.enforce_quotas(state)?;
+            profile::process(state, &self.inner.config);
+        });
+
+        if state.has_event() {
```

Comment:
This should be trivially true in `process_transaction`. Dynamic sampling can drop the event by returning an error. The same is true for all other process functions related to events.